labgrid/resource/docker.py

```diff
@@ -0,0 +1,127 @@
+    only one for each target.
+    """
+
+    def __attrs_post_init__(self):
+        super().__attrs_post_init__()
+        self.log = logging.getLogger('DockerManager')
+        self._client = dict()
+        self._docker_daemons_cleaned = list()
+
+    def on_resource_added(self, resource):
+        """
+        If the resource added is a DockerDaemon make sure this is the only one added for this target. If it is
+        create a docker client to be used to communicate with the docker daemon.
+        If the resource added is a DockerNetworkService validate a DockerDaemon exist for the target.
+        """
+        if isinstance(resource, DockerDaemon):
+            if resource.target.name in self._client:
+                raise EnvironmentError("Only one docker daemon is allowed for each target")
+            self._client[resource.target.name] = docker.DockerClient(base_url=resource.docker_daemon_url)
+            self._container_cleanup(self._client[resource.target.name])
+            resource.avail = True
+        else:
+            if not self._client[resource.target.name]:
+                raise EnvironmentError("Missing docker daemon resource")
+
+    def poll(self):
+        for resource in self.resources:
+            if not isinstance(resource, DockerDaemon):
+                resource.on_poll(self._client[resource.target.name])
+
+    def _container_cleanup(self, docker_client):
+        """
+        Fetches a list of containers from the daemon. Each container that has the label DOCKER_LG_CLEANUP_LABEL
+        is removed. This is for cleanup of potential dangling containers left over from earlier test runs.
+        """
+        if docker_client.api.base_url not in self._docker_daemons_cleaned:
+            container_list = docker_client.api.containers(
+                all=True, filters={"label": DockerConstants.DOCKER_LG_CLEANUP_LABEL})
+            for container in container_list:
+                if container['Labels'][DockerConstants.DOCKER_LG_CLEANUP_LABEL] == DockerConstants.DOCKER_LG_CLEANUP_TYPE_AUTO:
+                    self.log.info("Deleting container " + container['Names'][0])
+                    docker_client.api.remove_container(container['Id'], force=True)
+            self._docker_daemons_cleaned.append(docker_client.api.base_url)
+
+
+@target_factory.reg_resource
+@attr.s(cmp=False)
+class DockerDaemon(ManagedResource):
+    manager_cls = DockerManager
+    """ A ressource identifying a docker daemon """
+    docker_daemon_url = attr.ib(validator=attr.validators.instance_of(str))
+
+
+@target_factory.reg_resource
+@attr.s(cmp=False)
+class DockerNetworkService(ManagedResource, NetworkService):
+    """
+    The docker network service is a managed resource mapping a container name to a network service.
+    """
+    manager_cls = DockerManager
+    container_name = attr.ib(default=None, validator=attr.validators.instance_of(str))
+    address = attr.ib(default="", validator=attr.validators.in_([""]))
```

Comment:
Hmm, you're overriding the address attribute from `NetworkService` here.
I think this could work more cleanly by using the same approach we have in the `RemotePlaceManager`/`RemotePlace` and related resources:
You wouldn't define `DockerNetworkServices` explicitly, but the `DockerManager` would create `NetworkService` resources as needed and keep them up-to-date (see how `RemotePlaceManager` creates resources dynamically with `make_resource()`). This way, you could also keep the `avail` status of the `NetworkService` correct as the container is started and stopped.

Reply:
See comments from trxanga